editor/inserter/menu.js

```diff
@@ -307,16 +300,54 @@ export class InserterMenu extends Component {
 		);
 	}
 
+	renderBlocks( blocks, separatorSlug ) {
+		const { instanceId } = this.props;
+		const labelledBy = separatorSlug === undefined ? null : `editor-inserter__separator-${ separatorSlug }-${ instanceId }`;
+		return (
+			<div
+				className="editor-inserter__category-blocks"
+				tabIndex="0"
+				aria-labelledby={ labelledBy }
+			>
+				{ blocks.map( ( block ) => this.getBlockItem( block ) ) }
+			</div>
+		);
+	}
+
+	renderCategory( category, blocks ) {
+		const { instanceId } = this.props;
+		return blocks && (
+			<div key={ category.slug }>
+				<div
+					className="editor-inserter__separator"
+					id={ `editor-inserter__separator-${ category.slug }-${ instanceId }` }
+					aria-hidden="true"
+				>
+					{ category.title }
+				</div>
+				{ this.renderBlocks( blocks, category.slug ) }
+			</div>
+		);
+	}
+
+	renderCategories( visibleBlocksByCategory ) {
+		return getCategories().map(
```

Comment:
You're right on all accounts 👍  (To the final point, Lodash's `map` supports objects, returning an array, so that could have been an option)